File: sample-conference-videochat/sample-conference-videochat/Components/Calls/StreamParticipantViewController/StreamParticipantViewController.m
//
//  StreamParticipantViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 24.03.2021.
//  Copyright © 2021 Quickblox. All rights reserved.
//

#import "StreamParticipantViewController.h"
#import "StreamTitleView.h"

@interface StreamParticipantViewController ()
@property (strong, nonatomic) StreamTitleView *streamTitleView;
@end

@implementation StreamParticipantViewController

- (void)configureNavigationBarItems {
    self.streamTitleView = [[StreamTitleView alloc] init];
    self.navigationItem.titleView = self.streamTitleView;
    
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"menu_chat"] style:UIBarButtonItemStylePlain target:self action:@selector(didTapChat:)];
    self.navigationItem.leftBarButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.leftBarButtonItem.enabled = NO;
}

- (void)setupSession {
    // creating session
    NSString *conferenceID = self.conferenceSettings.conferenceInfo.conferenceID;
    self.session = [[QBRTCConferenceClient instance] createSessionWithChatDialogID:conferenceID conferenceType:QBRTCConferenceTypeVideo];
    if (!self.session) {
        return;
    }
    
    [self addToCollectionUserWithID:self.conferenceSettings.conferenceInfo.initiatorID];
    [self reloadContent];
//    self.localVideoView = nil;
}

- (void)setupLocalMediaStreamVideoCapture {
    // Listner cannot stream his video
    self.session.localMediaStream.videoTrack.videoCapture = nil;
}

- (void)setupAudioVideoEnabledCell:(ConferenceUserCell *)cell forUserID:(NSNumber *)userID {
    cell.videoEnabled = YES;
    cell.isMuted = YES;
}

- (void)configureToolBar {
    __weak __typeof(self)weakSelf = self;
    
    self.muteVideo = YES;
    self.muteAudio = YES;
    
    [self.toolbar addButton:[ButtonsFactory decline] action: ^(UIButton *sender) {
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf setupHideToolbarTimerWithTimeInterval:kHideTimeInterval];
        [strongSelf leaveFromRoomWithAnimated:YES completion:nil];
    }];
    
    [self.toolbar updateItems];
}

- (void)updateWithCreatedNewSession:(QBRTCConferenceSession *)session {
    self.navigationItem.leftBarButtonItem.enabled = YES;
    
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    
    QBRTCAudioSessionConfiguration *configuration = [[QBRTCAudioSessionConfiguration alloc] init];
    configuration.categoryOptions |= AVAudioSessionCategoryOptionDuckOthers;
    
    // adding blutetooth support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
    
    // adding airplay support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
    
    configuration.mode = AVAudioSessionModeVideoChat;
    
    [audioSession setConfiguration:configuration active:YES];
    
    session.localMediaStream.audioTrack.enabled = NO;
    session.localMediaStream.videoTrack.enabled = NO;
    
    __weak __typeof(self)weakSelf = self;
    [self.session listOnlineParticipantsWithCompletionBlock:^(NSArray<NSNumber *> * _Nonnull publishers, NSArray<NSNumber *> * _Nonnull listeners) {
        __typeof(weakSelf)strongSelf = weakSelf;
        for (NSNumber *userID in publishers) {
            if (userID.unsignedIntValue != self.conferenceSettings.conferenceInfo.initiatorID.unsignedIntValue) {
                continue;
            }
            [session subscribeToUserWithID:userID];
            break;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [strongSelf.streamTitleView setupStreamTitleViewOnLive:publishers.count > 0];
        });
    }];
}

- (void)session:(QBRTCConferenceSession *)session didReceiveNewPublisherWithUserID:(NSNumber *)userID {
    if (session != self.session) {
        return;
    }
    [self.session subscribeToUserWithID:userID];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.streamTitleView setupStreamTitleViewOnLive:YES];
        [self showControls:YES];
    });
    
    [self reloadContent];
}

- (void)removeUserFromCollection:(NSNumber *)userID {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.streamTitleView setupStreamTitleViewOnLive:NO];
        [self showControls:YES];
    });
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 1;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [self showControls:YES];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/StreamParticipantViewController/StreamParticipantViewController.swift
//
//  StreamParticipantViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 16.08.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC

class StreamParticipantViewController: BaseCallViewController {
    lazy private var streamTitleView: StreamTitleView = {
        let streamTitleView = StreamTitleView()
        return streamTitleView
    }()
    
    // MARK: Overrides
    override func setupSession() {
        // creating session
        session = QBRTCConferenceClient.instance().createSession(withChatDialogID: conferenceSettings.conferenceInfo.conferenceID,
                                                                 conferenceType: .video)
        
        guard session != nil else {
            return
        }

        addToCollectionUser(withID: NSNumber(value: conferenceSettings.conferenceInfo.initiatorID))
        reloadContent()
        localVideoView = nil
    }
    
    override func setupLocalMediaStreamVideoCapture() {
        // Listner cannot stream his video
        session?.localMediaStream.videoTrack.videoCapture = nil
    }
    
    override func configureNavigationBarItems() {
        
        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "menu_chat"),
                                                           style: .plain,
                                                           target: self,
                                                           action: #selector(didTapChat(_:)))
        navigationItem.leftBarButtonItem?.tintColor = .white
        navigationItem.leftBarButtonItem?.isEnabled = false
        navigationItem.rightBarButtonItem?.tintColor = .white
        navigationItem.rightBarButtonItem?.isEnabled = false
        
        navigationItem.titleView = streamTitleView
        navigationItem.rightBarButtonItem = nil
    }
    
    override func configureToolBar() {
        self.muteAudio = true
        self.muteVideo = true
        toolbar.add(ButtonsFactory.decline(), action: { [weak self] sender in
            guard let self = self else {
                return
            }
            self.setupHideToolbarTimerWithTimeInterval(BaseConstant.hideInterval)
            self.leaveFromCallAnimated(true)
        })
        toolbar.updateItems()
    }
    
    override func setupAudioVideoEnabledCell(_ cell: ConferenceUserCell, forUserID userID: UInt) {
        cell.videoEnabled = true
        cell.unMute = true
    }
    
    override func updateUIWithCreatedNewSession(_ session:QBRTCConferenceSession) {
        session.listOnlineParticipants(completionBlock: { [weak self] publishers, listeners in
            for userID in publishers {
                session.subscribeToUser(withID: userID)
            }
            DispatchQueue.main.async {
                self?.streamTitleView.setupStreamTitleViewOnLive(publishers.isEmpty == false)
            }
        })
    }
    
    override func session(_ session: QBRTCConferenceSession?, didReceiveNewPublisherWithUserID userID: NSNumber?) {
        guard let session = session, session == self.session, let userId = userID else {
            return
        }
            session.subscribeToUser(withID: userId)
            DispatchQueue.main.async {
                self.streamTitleView.setupStreamTitleViewOnLive(true)
                self.showControls(true)
            }
        reloadContent()
    }
    
    override func removeUserFromCollection(_ userID: NSNumber) {
        
        DispatchQueue.main.async {
            self.streamTitleView.setupStreamTitleViewOnLive(false)
            self.showControls(true)
        }
    }
    
    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 1
    }
    
    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        self.showControls(true)
    }
    
    override func userView(userID: UInt) -> UIView? {
        if let remoteVideoTraсk = session?.remoteVideoTrack(withUserID: NSNumber(value: userID)) {
            //Opponents
            if let remoteVideoView = videoViews[userID] as? QBRTCRemoteVideoView {
                return remoteVideoView
            } else {
                let remoteVideoView = QBRTCRemoteVideoView(frame: CGRect(x: 2.0, y: 2.0, width: 2.0, height: 2.0))
                remoteVideoView.videoGravity = AVLayerVideoGravity.resizeAspect.rawValue
                videoViews[userID] = remoteVideoView
                remoteVideoView.setVideoTrack(remoteVideoTraсk)
                
                return remoteVideoView
            }
    }
        return nil
    }
}
